ssl/t1_enc.c

```diff
@@ -379,12 +419,33 @@ int tls1_change_cipher_state(SSL *s, int which)
            TLS_CIPHER_AES_GCM_128_IV_SIZE);
     memcpy(crypto_info.salt, geniv, TLS_CIPHER_AES_GCM_128_SALT_SIZE);
     memcpy(crypto_info.key, key, EVP_CIPHER_key_length(c));
-    memcpy(crypto_info.rec_seq, &s->rlayer.write_sequence,
-           TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+    if (which & SSL3_CC_WRITE)
+        memcpy(crypto_info.rec_seq, &s->rlayer.write_sequence,
+                TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+    else
+        memcpy(crypto_info.rec_seq, &s->rlayer.read_sequence,
+                TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE);
+
+    if (which & SSL3_CC_READ) {
+        count_unprocessed = count_unprocessed_records(s);
+        if (count_unprocessed < 0)
+            goto skip_ktls;
+
+        /* increment the crypto_info record sequence */
+        while (count_unprocessed) {
+            for (bit = 7; bit >= 0; bit--) { /* increment */
+                ++crypto_info.rec_seq[bit];
+                if (crypto_info.rec_seq[bit] != 0)
+                    break;
+            }
+            count_unprocessed--;
+        }
+    }
```

Comment:
It looks to me like you are assuming that the read buffer only contains whole records, not partial ones. I don't think you can make that assumption.

Reply:
There are two scenarios to consider, i.e. whether or not read_ahead is set.

If read_ahead is not set (the default) then `ssl3_get_record` will fill the read buffer in two steps. First it reads a whole record header (5 bytes). Then from the record header it works out the length of the record it is expecting and reads that many bytes. Once its got the required number of bytes it processes the record and we continue to process the contained data. If that data holds a CCS message then we will end up in this function. Since `ssl3_get_record` only ever reads just the right amount of data (in the case where read_ahead is not set) then the read buffer should never have any remaining data left in it.

If read_ahead is set then `ssl3_get_record` works the same way except when it attempts to do a read from the underlying BIO it will read as much data as is available into the read buffer. This may be more than it actually needs to process a whole record. Therefore when `ssl3_get_record` finishes there may be left over data in the read buffer. This is unprocessed data so we have no idea whether it consists of whole or partial records.

Perhaps the simplest way to handle this is to not allow KTLS to engage on the receive side if read_ahead is set. In that case I don't think you need this code for counting unprocessed records at all (because we should never have unprocessed data). OTOH if you do want KTLS to engage even if read_ahead is set then you will have to handle partial records.

Reply:
Thanks.

The current code will fail to start ktls if there is a partially read record. (`count_unprocessed_records` returns -1). If read_ahead is enabled this might succeed or fail, which is a valid but somewhat unexpected behavior.

I do want ktls to work properly regardless of the previous data read by read_ahead. But, I think this could be done in a follow-up patch, right?

Do you think that it would be possible to use the `read_ahead` functionality only after the handshake is complete?

Reply:
Alternatively, we could use `read_ahead` with a buffer that could contain up to an additional foull record (~16KB), which would be reserved during the handshake phase. The code above would use this reserved buffer to read the remaining portion of the partial record. Finally, this buffer would be used by `read_ahead` for reading further incoming data.

Reply:
> I do want ktls to work properly regardless of the previous data read by read_ahead. But, I think this could be done in a follow-up patch, right?

Yes - a follow up patch would be fine. Although I suggest if we got that path then some interim comment in the code might be useful explaining the interaction with read_ahead as it stands now.

> Do you think that it would be possible to use the `read_ahead` functionality only after the handshake is complete?

Well a user application can set it at any time. I'm not sure we would want to enforce that read_ahead only happens after the handshake is complete.

Reply:
I've added a comment there.